### ISSUE TITLE
Use BLOB type for serializing params for UDF function

### ISSUE THREAD

**Issue by bpaquet** [aerospike/aerospike-client-php#40]
Is there a way to always use blob serialization when calling an UDF function ?

**Comment by rbotzer** [comment]
Yes, you can create pack bytes into a PHP string, pass that as the argument to the record UDF, cast it to bytes(), then inspect bytes as meaningful params. You'll use API methods in Aerospike's [bytes Lua module](http://www.aerospike.com/docs/udf/api/bytes.html).

Assume you have registered the following UDF module as `example_bytes_udf`:
```lua
function inspect_bytes_input(rec, input)
    local bytes_input = bytes(32)
    bytes.append_string(bytes_input, input)
    local ret = map()
    ret['size'] = bytes.size(bytes_input)
    ret['type'] = bytes.get_type(bytes_input)
    ret['byte1'] = bytes.get_byte(bytes_input, 1)
    ret['byte2'] = bytes.get_byte(bytes_input, 2)
    ret['byte3'] = bytes.get_byte(bytes_input, 3)
    ret['byte4'] = bytes.get_byte(bytes_input, 4)
    return ret
end
```

I can call this from PHP as follows:
```php
$db->apply($key, "example_bytes_udf", "inspect_bytes_input", array("moo\7x\7"), $ret_val);
var_dump($ret_val);
```

And I get the following:
```
array(6) {
  ["size"]=>
  int(6)
  ["byte1"]=>
  int(109)
  ["byte2"]=>
  int(111)
  ["byte3"]=>
  int(111)
  ["byte4"]=>
  int(7)
  ["type"]=>
  int(4)
}
```

Which means it sees 6 chars in the input (two of them are the \7 bell character), inspecting individual bytes shows the expected values in each position, and the input is an as_bytes of type 'generic'.

I hope that helps.

**Comment by bpaquet** [comment]
Mmh, my question were not very good.

So.
I'm using aerospike as redis drop in. Goal : no modification in client code.
For that, I'have written this small wrapper
https://github.com/bpaquet/aerospike_redis_php

The only problem I have is for storing gzcompress string into aerospike.
With redis, I can do $r->set('toto', gzcompress(big_json));
With aerospike too, bit the string is truncated at first NULL.

I have read the 'known problem' of the readme, but custom serializer code
seems to not be used with UDF.
I have try to unpack('C*') the string to force the drive to use bytes
array, but it perform very slowly, and I get  some memory overflow.

The only workaround I have is to base64 encode the gzcompress string :(

Any better idea ?

**Comment by rbotzer** [comment]
The solution would be for us to ensure UDF behaves in a similar way to the write operations. With the bit of Redis you've mentioned you're doing custom serialization anyway. Once UDFs behave as the put() would with a custom session handler, you have more or less drop-in code.

However, I'm not sure why you'd want to pass serialized PHP to a record-UDF (apply() ). In your example you're just writing to a record. That's what put() and operate() are for in Aerospike. What kind of work on the serialized PHP would you need to do?

**Comment by bpaquet** [comment]
On Tue, Apr 14, 2015 at 9:43 AM, Ronen Botzer <notifications@github.com>
wrote:

> The solution would be for us to ensure UDF behaves in a similar way to the
> write operations. With the bit of Redis you've mentioned you're doing
> custom serialization anyway. Once UDFs behave as the put() would with a
> custom session handler, you have more or less drop-in code.
>
> However, I'm not sure why you'd want to pass serialized PHP to a
> record-UDF (apply() ). In your example you're just writing to a record.
> That's what put() and operate() are for in Aerospike. What kind of work on
> the serialized PHP would you need to do?
>

No, I do not want pass serialized PHP. I want to pass binaray string

$redis-set(gzcompress(json_encode(big_object));

I do not want to change gzcompress(json_encode(big_object)) because it's
not on my scope : I do not want to change the app code.

So, the default serialization for udf use String, not bytes.
Is there any way to force it ?
I dit not success to use a custom serializer.

**Comment by rbotzer** [comment]
Bertrand, json-encoding, then compressing your object _is_ a form of custom serialization. by comparison, the PHP serialize() function is one of the native ones in PHP's 'standard library'. To register your own serializer you'd simply wrap gzcompress(json_encode($val)) in a function then register its name. I wasn't suggesting you change your app code. What I was pointing out earlier was that right now the apply() function doesn't respect a registered serializer, but instead uses the default PHP mechanism. That's inconsistent with put() so we will change it.

Right now UDFs can take an array of arguments. You can pack binary into any string and pass it as an argument, as long as it doesn't have a \0 in the middle of it. When I fix apply() to enable custom serialization that workaround you saw in the README will work on it as well.

But really, why do you need a record UDF (i.e. apply())? If you are trying to set a bin in a record you should use the put() method. :question:

**Comment by bpaquet** [comment]
I can use put for redis get / set operations.

But for others like rpush, lpop and so one, I have to use UDF.

I'm waiting for your patch :)

**Comment by rbotzer** [comment]
Alrighty. I opened a couple of JIRA tickets over here. Hope to get this resolved sooner rather than later. I'll update you.
